x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -421,6 +436,35 @@ public Entry<CacheFileRegion> get(KeyType cacheKey, long fileLength, int region)
         return entry;
     }
 
+    /**
+     * Fetch and cache the full blob for the given cache entry from the remote repository
+     * @param cacheKey  the key to fetch data for
+     * @param length    the length of the blob to fetch
+     * @param writer    a writer that handles writing of newly downloaded data to the shared cache
+     */
+    public void fetchFullEntry(KeyType cacheKey, long length, RangeMissingHandler writer) {
```

Comment:
There's actually one problem with the `writer` approach and the way we currently plan to use this. If we cancel inside the `writer`, then we will still acquire all the pages for a file we want to write. I think we need cancellation to live up the stack in this method's API?

Reply:
The pages all get acquired in the main thread before any downloading happens, so I'm not sure that implementing cancellation here will help.  Maybe the ActionListener passed to `entry.chunk.populateAndRead()` should evict the page if there's an Exception thrown during the download?

Reply:
Hm, except that ActionListener doesn't actually get called for write fails - I think I need a second listener here

Reply:
Ah you're right, sorry I had this a little backwards. It's a little weird for it to work this way I guess? We could be acquiring lots of pages here, but those then get evicted before anything can be done with them. But lets leave that for later :)

Reply:
Have updated this so that we evict the page if there's an error during the write.